include/quic/parser.hpp

```diff
@@ -0,0 +1,421 @@
+            data += ':';
+            data.append(d);
+
+            req_time = get_time();
+            timeout = req_time + TIMEOUT;
+        }
+
+        bool is_expired(std::chrono::steady_clock::time_point tp) { return timeout < tp; }
+
+        message to_message(bool timed_out = false) { return {return_sender, data, timed_out}; }
+
+        std::string_view view() { return {data}; }
+        std::string&& payload() { return std::move(data); }
+    };
+
+    class bparser : public Stream
+    {
+      private:
+        // outgoing requests awaiting response
+        std::deque<std::shared_ptr<sent_request>> sent_reqs;
+
+        std::string buf;
+        std::string size_buf;
+
+        size_t current_len{0};
+
+        std::atomic<int64_t> next_rid{0};
+
+        friend class sent_request;
+        std::function<void(message)> recv_callback;
+
+      public:
+        template <typename... Opt>
+        explicit bparser(Connection& _c, Endpoint& _e, Opt&&... opts) : Stream{_c, _e}
+        {
+            ((void)handle_bp_opt(std::forward<Opt>(opts)), ...);
+        }
+
+        ~bparser() { sent_reqs.clear(); }
+
+        std::weak_ptr<bparser> weak_from_this() { return std::dynamic_pointer_cast<bparser>(shared_from_this()); }
+
+        void request(std::string endpoint, std::string body)
+        {
+            log::debug(bp_cat, "{} called", __PRETTY_FUNCTION__);
+
+            auto req = make_request(std::move(endpoint), std::move(body));
+            send(req->view());
+
+            sent_reqs.push_back(std::move(req));
+        }
+
+        void command(std::string endpoint, std::string body)
+        {
+            log::debug(bp_cat, "{} called", __PRETTY_FUNCTION__);
+
+            auto req = make_command(std::move(endpoint), std::move(body));
+            send(req->payload());
+        }
+
+        void respond(int64_t rid, std::string body, bool error = false)
+        {
+            log::debug(bp_cat, "{} called", __PRETTY_FUNCTION__);
+
+            auto req = make_response(rid, std::move(body), error);
+            send(req->payload());
+        }
+
+        void check_timeouts()
+        {
+            const auto& now = get_time();
+
+            do
+            {
+                auto& f = sent_reqs.front();
+
+                if (f->is_expired(now))
+                {
+                    recv_callback(f->to_message(true));
+                    sent_reqs.pop_front();
+                }
+                else
+                    return;
+
+            } while (not sent_reqs.empty());
+        }
+
+        void receive(bstring_view data) override
+        {
+            log::info(bp_cat, "bparser recv data callback called!");
+            log::debug(bp_cat, "Received data: {}", buffer_printer{data});
+
+            try
+            {
+                process_incoming(to_sv(data));
+            }
+            catch (std::exception& e)
+            {
+                log::error(bp_cat, "Exception caught: {}", e.what());
```

Comment:
We need to close the stream here, because if we hit an exception then we're in an unrecoverable state on the stream and we really can't meaningfully keep parsing.